pkg/testing/runner/runner.go

```diff
@@ -490,288 +562,170 @@ func (r *Runner) setupCloud(ctx context.Context) error {
+			})
 		}
-		go func(ctx context.Context, version string, resp *essCloudResponse) {
-			ctx, cancel := context.WithTimeout(ctx, 10*time.Minute)
-			defer cancel()
-			ready, err := essClient.DeploymentIsReady(ctx, resp.resp.ID, 30*time.Second)
+	}
+	if len(requests) == 0 {
+		// no need to request any other stacks
+		r.stacksReady.Done()
+		return nil
+	}
+
+	// start go routine to provision the needed stacks
+	go func(ctx context.Context) {
+		defer r.stacksReady.Done()
+
+		ctx, cancel := context.WithTimeout(ctx, 10*time.Minute)
+		defer cancel()
+
+		stacks, err := r.sp.Provision(ctx, requests)
+		if err != nil {
+			r.stacksErr = err
+			return
+		}
+		for _, stack := range stacks {
+			err := r.addOrUpdateStack(stack)
 			if err != nil {
-				r.logger.Logf("Failed to check for cloud %s to be ready: %s", version, err)
+				r.stacksErr = err
+				return
 			}
-			resp.subMx.RLock()
-			subs := make([]chan *ess.CreateDeploymentResponse, len(resp.subCh))
-			copy(subs, resp.subCh)
-			resp.subCh = make([]chan *ess.CreateDeploymentResponse, 0)
-			resp.done = true
-			resp.ready = ready
-			resp.subMx.RUnlock()
-			var send *ess.CreateDeploymentResponse
-			if ready {
-				send = resp.resp
-			}
-			for _, sub := range subs {
-				select {
-				case <-ctx.Done():
-					return
-				case sub <- send:
+			for batchID, batchVersion := range batchToVersion {
+				if batchVersion == stack.Version {
+					r.batchToStack[batchID] = stack
 				}
 			}
-		}(ctx, version, essResp)
-	}
+		}
+	}(ctx)
+
 	return nil
 }
 
-func (r *Runner) cleanupCloud() {
-	r.batchToCloudMx.Lock()
-	defer r.batchToCloudMx.Unlock()
-
-	essClient := ess.NewClient(ess.Config{
-		ApiKey: r.cfg.ESS.APIKey,
-	})
+func (r *Runner) getStackForBatchID(id string) (Stack, bool) {
+	stack, ok := r.batchToStack[id]
+	return stack, ok
+}
 
-	var removed []string
-	for _, cloud := range r.batchToCloud {
-		if slices.Contains(removed, cloud.resp.ID) {
-			// already removed
-			continue
-		}
-		removed = append(removed, cloud.resp.ID)
-		err := func() error {
-			ctx, cancel := context.WithTimeout(context.Background(), 2*time.Minute)
-			defer cancel()
-			return essClient.ShutdownDeployment(ctx, cloud.resp.ID)
-		}()
-		if err != nil {
-			r.logger.Logf("Failed to cleanup cloud %s: %s", cloud.resp.ID, err)
+func (r *Runner) findInstance(id string) (StateInstance, bool) {
+	r.stateMx.Lock()
+	defer r.stateMx.Unlock()
+	for _, existing := range r.state.Instances {
+		if existing.ID == id {
+			return existing, true
 		}
 	}
-
-	r.batchToCloud = make(map[string]*essCloudResponse)
+	return StateInstance{}, false
 }
 
-func (r *Runner) getCloudForBatchID(id string) (chan *ess.CreateDeploymentResponse, error) {
-	r.batchToCloudMx.RLock()
-	essResp, ok := r.batchToCloud[id]
-	if !ok {
-		r.batchToCloudMx.RUnlock()
-		return nil, fmt.Errorf("no batch with ID %s", id)
-	}
-	r.batchToCloudMx.RUnlock()
+func (r *Runner) addOrUpdateInstance(instance StateInstance) error {
+	r.stateMx.Lock()
```

Comment:
are we calling `find*` and `add*` functions from different goroutines? I see a lot of mutex lock/unlock for state...

Reply:
Yes, controlling instances is completely parallelized.